instrumentation/rails/lib/opentelemetry/instrumentation/rails/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module Rails
+      module Middlewares
+        # TracerMiddleware propagates context and instruments Rails requests
+        # by way of its middleware system
+        class TracerMiddleware
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env)
+            extracted_context = OpenTelemetry.propagation.http.extract(env)
+            OpenTelemetry::Context.with_current(extracted_context) do
+              tracer.in_span(env['PATH_INFO'], attributes: request_span_attributes(env: env), kind: :server) do |request_span|
+                @app.call(env).tap do |status, headers, response|
+                  set_attributes_after_request(
+                    request_span,
+                    env: env,
+                    status: status,
+                    headers: headers,
+                    response: response
+                  )
+                end
+              end
+            end
+          ensure
+            OpenTelemetry::Trace.current_span(extracted_context).finish if extracted_context
+          end
+
+          private
+
+          def request_destination_info(env)
+            ::Rails.application.routes.router.recognize(ActionDispatch::Request.new(env)) { |_route, params| return params }
```

Comment:
So I'm not sure if I should use this or not, we can actually infer the controller and action in this middleware **before** `@app.call` is made.  In the event that something goes wrong in a middleware we could use that information to normalize the operation name, but I don't know if I should consider this a cheap enough operation to do it on every request.

Reply:
I feel like that should be a fallback rather than the default, but I don't really know the cost of the operation.

Does the `_route` contain any useful info, or just the `params`?

Reply:
The route object contains a lot of info, I spent some time digging in it and the most useful thing IMO was the `requirements` method which == params.

#### For our example controller
```rb
Route = #<ActionDispatch::Journey::Route:0x00007fa1f0719eb0>
Params = {:controller=>"example", :action=>"ok"}
```

#### If you hit the built in "welcome to rails" page
```rb
Route = []
# We get the namespace as well
Params = {:controller=>"rails/welcome", :action=>"index"}
```

The route object details can be found here https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/journey/route.rb

Note that params is just [`route.requirements`](https://github.com/rails/rails/blob/master/actionpack/lib/action_dispatch/journey/route.rb#L91-L103).

Reply:
From the code you pointed at:
```
      #   get 'photo/:id', :controller => 'photos', :action => 'show',
```
I think the spec would like us to use `photo/:id` as the span name, but I'm not entirely sure how to get that from the Route.